cppapi/client/zmqeventconsumer.cpp

```diff
@@ -1294,21 +1294,29 @@ void ZmqEventConsumer::connect_event_channel(string &channel_name,TANGO_UNUSED(D
 // If the server has returned several possible ZMQ endpoints (because several NIC boards on server host), check which
 // one is correct
 //
+    size_t nb_endpoints = ev_svr_data->svalue.length();
 
-    size_t nb_endpoint = ev_svr_data->svalue.length();
-    nb_endpoint = nb_endpoint >> 1;
+    Tango::DevLong server_tango_lib_version = ev_svr_data->lvalue[0];
+    if(server_tango_lib_version >= 930)
+    {
+        // ZmqEventSubscriptionChange returns the ZMQ message filters used for the attribute
+        // and for the heartbeat events at the end of the string array part of the returned DevVarLongStringArray
+        nb_endpoints -= 2;
+    }
+
+    nb_endpoints = nb_endpoints >> 1;
     size_t valid_endpoint = 0;
 
-    if (nb_endpoint != 1)
+    if (nb_endpoints != 1)
     {
-        for (valid_endpoint = 0;valid_endpoint < nb_endpoint;valid_endpoint++)
+        for (valid_endpoint = 0; valid_endpoint < nb_endpoints; valid_endpoint++)
         {
             string endpoint(ev_svr_data->svalue[valid_endpoint << 1]);
             if (check_zmq_endpoint(endpoint) == true)
                 break;
```

Comment:
In the scenario you described in the original post, was channel name (last element on the list) used as zmq endpoint?
If yes, then:
1. It means that no other endpoint on the list passes the `check_zmq_endpoint` check- client won't be able to receive events anyway.
2. Maybe the `check_zmq_endpoint` should be improved to reject invalid values? I think channel name starts with 'tango://', right?

Reply:
> In the scenario you described in the original post, was channel name (last element on the list) used as zmq endpoint?

Yes

> If yes, then:
> 
>     1. It means that no other endpoint on the list passes the `check_zmq_endpoint` check- client won't be able to receive events anyway.

Indeed

>     2. Maybe the `check_zmq_endpoint` should be improved to reject invalid values? I think channel name starts with 'tango://', right?

The problem is that we pass values which are not endpoints. My proposal is to avoid that.
We have a way to know whether the 2 last values are endpoints or not thanks to the remote server tango lib version also returned as result of ZmqEventSubscriptionChange command.

If we improve check_zmq_endpoint, we could improve it so endpoint containing hostnames would be handled properly, because as it is now, check_zmq_endpoint is expecting an IP address in the given end point.